denonavr/api.py

```diff
@@ -316,134 +318,210 @@ def is_default_async_client(self) -> bool:
-                _LOGGER.error(
-                    "Unexpected exception on telnet reconnect", exc_info=True)
-            else:
-                _LOGGER.info("Telnet reconnected")
-                self._healthy = True
-                self._monitor_task = asyncio.create_task(self._async_monitor())
-                return
-
-            await asyncio.sleep(backoff)
-            backoff = min(30.0, backoff*2)
 
-    async def _async_monitor(self):
-        """Read the messages on the TCP socket."""
-        data = bytearray()
-        while not self._reader.at_eof():
-            try:
-                chunk = await asyncio.wait_for(
-                    self._reader.read(_SOCKET_READ_SIZE), 30.0)
-            except (asyncio.TimeoutError, IOError, OSError):
-                _LOGGER.info(
-                    "Lost telnet connection to receiver, reconnecting")
-                self._monitor_task = asyncio.create_task(
-                    self._async_reconnect())
-                return
-            except asyncio.CancelledError:
-                _LOGGER.debug("Stopped telnet monitoring")
-                return
-            except Exception:    # pylint: disable=broad-except
-                _LOGGER.error(
-                    "Unexpected exception while monitoring telnet",
-                    exc_info=True)
-                return
-            # pylint: disable=consider-using-enumerate
-            for i in range(0, len(chunk)):
-                # Messages are CR terminated
-                if chunk[i] != 13:
-                    data += chunk[i].to_bytes(1, byteorder='big')
+        async with self._connect_lock:
+            while self._connection_enabled is True and not self.healthy:
+                try:
+                    await self._async_establish_connection()
+                except AvrTimoutError:
+                    _LOGGER.debug(
+                        "%s: Timeout exception on telnet reconnect",
+                        self.host
+                    )
+                except AvrNetworkError as ex:
+                    _LOGGER.debug("%s: %s", self.host, ex, exc_info=True)
+                except Exception:    # pylint: disable=broad-except
+                    _LOGGER.error(
+                        "%s: Unexpected exception on telnet reconnect",
+                        self.host,
+                        exc_info=True
+                    )
                 else:
-                    await self._process_event(str(data, 'utf-8'))
-                    data = bytearray()
+                    _LOGGER.info("%s: Telnet reconnected", self.host)
+                    return
 
-        if self.connected is True:
-            _LOGGER.info(
-                "Telnet connection terminated by receiver, reconnecting")
-            self._monitor_task = asyncio.create_task(self._async_reconnect())
+                await asyncio.sleep(backoff)
+                backoff = min(30.0, backoff*2)
```

Comment:
This should be outside the lock. See previous comment.
```suggestion
            await asyncio.sleep(backoff)
            backoff = min(30.0, backoff*2)
```

Reply:
Resolved in https://github.com/dcmeglio/denonavr/commit/c4ef1fcd87ed5c51313699dde6f5d6ca46c40a2b